### PR TITLE
Add a toThrowWithMessage matcher

### PR DESCRIPTION
**Pull request by natealcedo**
### What
This pull request adds in a new matcher for asserting that a callback function throws an error of a given type and given error message.

### Why
This feature has been requested in core but after some discussion, it was decided that implementing it here would be the better solution. 

This PR fixes the following issues
1. https://github.com/facebook/jest/issues/7000
2. https://github.com/facebook/jest/issues/3659
3. #157 

### Notes
I've gone down the road of testing the matcher both by extending expect with this matcher and by invoking the matcher itself. Since this repository doesn't have some of the util tools that the Jest repo has, namely [runJest](https://github.com/facebook/jest/blob/master/e2e/runJest.js), I am not able to test the failing scenarios since I can't capture whatever is logged to stdErr to assert that the error message is indeed correct. As such, I've just invoked the matcher itself and asserted whether it passes or not and used snapshot testing for asserting error messages.

Also, the goal for me in this matcher was to stick as close to the default `toThrowError`matcher. As such, this matcher also accepts both a `string` and `regex` for matching the error message. 

The code is abit verbose right now but I'm more than happy to have a discussion on how to amend it moving forward.
### Housekeeping

- [x] Unit tests
- [x] Documentation is up to date
- [x] No additional lint warnings
- [x] Add yourself to contributors list (`yarn contributor`)
- [x] [Typescript definitions](https://github.com/jest-community/jest-extended/blob/master/types/index.d.ts) are added/updated where relevant

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -398,6 +399,30 @@ test('passes when value is a function', () => {
 });
 ```
 
+#### .toThrowWithMessage(type, message)
+
+Use `.toThrowWithMessage` when checking if a callback function throws an error with a given error type and given error message.
```
Comment (mattphillips) [inline README.md]: Perhaps it's worth mentioning that the message can be either a `String` or `RegExp`?

File: types/index.d.ts [hunk 2]
```diff
@@ -337,5 +337,13 @@ declare namespace jest {
...
      */
     toIncludeMultiple(substring: string[]): R;
+
+    /**
+     * Use `.toThrowWithMessage` when checking if a callback function throws an error of a given type with a given error message.
+     *
+     * @param {Function} type
+     * @param {String | RegExp} message
+     */
+    toThrowWithMessage(type: Function, message: string | RegExp): R;
```
Comment (mattphillips) [inline types/index.d.ts]: This needs adding to the `expect` interface too

File: src/matchers/toThrowWithMessage/index.js [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+Thrown:
+${printReceived(received)}
+`;
+
+export default {
+  toThrowWithMessage: (callback, type, message) => {
+    if (!callback || typeof callback !== 'function') {
+      return {
+        pass: false,
+        message: () => `
```
Comment (mattphillips) [inline src/matchers/toThrowWithMessage/index.js]: The spacing of these validation errors is tabbed in too much for the printouts (see the snapshots)

**Comment by mattphillips** [comment]
@natealcedo this is looking good just a few small comments :)

File: src/matchers/toThrowWithMessage/index.js (hunk 3 above)
Comment (natealcedo) [reply src/matchers/toThrowWithMessage/index.js]: Alright, I've pushed in a change for this. Do take a look.

File: types/index.d.ts (hunk 2 above)
Comment (mattphillips) [reply types/index.d.ts]: Ah we just merged the types for asymmetric matchers so needed to update this file twice - all good though I just did it for you :)

**Comment by mattphillips** [comment]
Thanks @natealcedo 👍

**Pull request merged**